python/ray/_raylet.pyx

```diff
@@ -937,99 +925,266 @@ cdef execute_streaming_generator(
-            break
-        except StopIteration:
-            break
+            output_or_exception = next(gen)
         except Exception as e:
-            create_generator_error_object(
-                e,
-                worker,
-                task_type,
-                caller_address,
-                task_id,
-                serialized_retry_exception_allowlist,
-                function_name,
-                function_descriptor,
-                title,
-                actor,
-                actor_id,
-                return_size,
-                generator_index,
-                is_async,
-                &return_obj,
-                is_retryable_error,
-                application_error
-            )
+            output_or_exception = e
+        done = report_streaming_generator_output(output_or_exception, context)
+        if done:
+            break
 
-            streaming_generator_returns[0].push_back(
-                c_pair[CObjectID, c_bool](
-                    return_obj.first,
-                    is_plasma_object(return_obj.second)))
 
-            CCoreWorkerProcess.GetCoreWorker().ReportGeneratorItemReturns(
-                return_obj,
-                generator_id,
-                caller_address,
-                generator_index,
-                attempt_number)
-            generator_index += 1
-            break
-        else:
-            # Report the intermediate result if there was no error.
-            create_generator_return_obj(
-                output,
-                generator_id,
-                worker,
-                caller_address,
-                task_id,
-                return_size,
-                generator_index,
-                is_async,
-                &return_obj)
-            # Del output here so that we can GC the memory
-            # usage asap.
-            del output
-
-            streaming_generator_returns[0].push_back(
-                c_pair[CObjectID, c_bool](
-                    return_obj.first,
-                    is_plasma_object(return_obj.second)))
+async def execute_streaming_generator_async(
+        context: StreamingGeneratorExecutionContext):
+    """Execute a given generator and streaming-report the
+        result to the given caller_address.
 
-            logger.debug(
-                "Writes to a ObjectRefStream of an "
-                "index {}".format(generator_index))
-            CCoreWorkerProcess.GetCoreWorker().ReportGeneratorItemReturns(
-                return_obj,
-                generator_id,
-                caller_address,
-                generator_index,
-                attempt_number)
-            generator_index += 1
+    This method is same as `execute_streaming_generator_sync`,
+    but it should be used inside an async event loop.
+
+    NOTE: when this function is run inside an event loop,
+    some of core worker APIs could be executed inside
+    a separate thread (because we run event loop in a differnt).
```

Comment:
```suggestion
    NOTE: since this function runs inside an event loop thread,
    some of core worker APIs will be executed inside
    the event loop thread as well.
```